contrib/inventory/foreman.py

```diff
@@ -0,0 +1,394 @@
+#!/usr/bin/env python
+# vim: set fileencoding=utf-8 :
+#
+# Copyright (C) 2016 Guido Günther <agx@sigxcpu.org>,
+#                    Daniel Lobato Garcia <dlobatog@redhat.com>
+#
+# This script is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# Ansible is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with it.  If not, see <http://www.gnu.org/licenses/>.
+#
+# This is somewhat based on cobbler inventory
+
+# Stdlib imports
+# __future__ imports must occur at the beginning of file
+from __future__ import print_function
+try:
+    # Python 2 version
+    import ConfigParser
+except ImportError:
+    # Python 3 version
+    import configparser as ConfigParser
+import json
+import argparse
+import copy
+import os
+import re
+import sys
+from time import time
+from distutils.version import LooseVersion, StrictVersion
+
+# 3rd party imports
+import requests
+if LooseVersion(requests.__version__) < LooseVersion('1.1.0'):
+    print('This script requires python-requests 1.1 as a minimum version')
+    sys.exit(1)
+
+from requests.auth import HTTPBasicAuth
+
+class ForemanInventory(object):
+
+    def __init__(self):
+        self.inventory = dict()  # A list of groups and the hosts in that group
+        self.cache = dict()   # Details about hosts in the inventory
+        self.params = dict()  # Params of each host
+        self.facts = dict()   # Facts of each host
+        self.hostgroups = dict()  # host groups
+        self.config_paths = [
+            "/etc/ansible/foreman.ini",
+            os.path.dirname(os.path.realpath(__file__)) + '/foreman.ini',
+        ]
+        env_value = os.environ.get('FOREMAN_INI_PATH')
+        if env_value is not None:
+            self.config_paths.append(os.path.expanduser(os.path.expandvars(env_value)))
+
+    @staticmethod
+    def to_safe(word):
+        '''Converts 'bad' characters in a string to underscores
+
+        so they can be used as Ansible groups
+
+        >>> ForemanInventory.to_safe("foo-bar baz")
+        'foo_barbaz'
+        '''
+        regex = "[^A-Za-z0-9\_]"
+        return re.sub(regex, "_", word.replace(" ", ""))
+
+    def json_format_dict(self, data, pretty=False):
+        """Converts a dict to a JSON object and dumps it as a formatted string"""
+
+        if pretty:
+            return json.dumps(data, sort_keys=True, indent=2)
+        else:
+            return json.dumps(data)
+
+    def push(self, d, k, v):
```

Comment:
Actually... Even better: use a defaultdict instead of this.  This is how defaultdicts work:
``` python
from collections import defaultdict

foo = defaultdict(list)
foo[k].append(v)
```